### PR TITLE
Remove bjerrum length n from long range interactions

### LINKED ISSUE #1382: Remove dependenceies from thermostat e.g. in long range interactions
Remove bjerrum length

### PR DESCRIPTION
**Pull request by RudolfWeeber**
Fixes #1382 

This makes the long range interactions independent of the thermostat, by switching from the bjerrum length as an argument to the Coulomb constant
C=1/(4pi epsilon_0 epsilon_r) =l_B *kT

This also gets rid of the effect that the meaning of the Bjerrum length silently changes to the Coulomb constant, when the temperature is set to zero, and back to Bjerrum length, when kT>0 again.

This PR updates the core and interface for all electrostatic and magnetostatic interactions and electrokinetics as well as the test suite and samples (in so far as they are running)
Also done are the docs of those interactions that have up-to-date python-based docs.
The EK docs (currently being converted to Py by @rempferg ) are untouched.
The tutorials are not updated. I suggest to defer those, until it is clear if the IPython-based ones are there in time for the release.

### CODE REVIEW

File: src/python/espressomd/magnetostatics.pyx [hunk 1]
```diff
@@ -166,7 +132,7 @@ IF DP3M == 1:
                     "mesh_off should be a list of length 3 and values between 0.0 and 1.0")
 
         def valid_keys(self):
-            return "prefactor", "alpha_L", "r_cut_iL", "mesh", "mesh_off", "cao", "inter", "accuracy", "epsilon", "cao_cut", "a", "ai", "alpha", "r_cut", "inter2", "cao3", "additional_mesh", "bjerrum_length", "tune"
+            return "prefactor", "alpha_L", "r_cut_iL", "mesh", "mesh_off", "cao", "inter", "accuracy", "epsilon", "cao_cut", "a", "ai", "alpha", "r_cut", "inter2", "cao3", "additional_mesh", "prefactor", "tune"
```
Comment (KonradBreitsprecher) [inline src/python/espressomd/magnetostatics.pyx]: prefacor appears twice in valid_keys

File: src/python/espressomd/magnetostatics.pyx [hunk 2]
```diff
@@ -371,7 +337,7 @@ IF DIPOLES == 1:
                 return ()
     
             def valid_keys(self):
-                return ("bjerrum_length", "prefactor")
+                return ("prefactor", "prefactor")
```
Comment (KonradBreitsprecher) [inline src/python/espressomd/magnetostatics.pyx]: prefactor appears twice

File: src/python/espressomd/magnetostatics.pyx [hunk 3]
```diff
@@ -277,7 +243,7 @@ IF DIPOLES == 1:
             return ()
 
         def valid_keys(self):
-            return ("bjerrum_length", "prefactor")
+            return ("prefactor", "prefactor")
```
Comment (KonradBreitsprecher) [inline src/python/espressomd/magnetostatics.pyx]: prefactor appears twice in valid_keys

File: src/python/espressomd/magnetostatics.pyx [hunk 4]
```diff
@@ -312,7 +278,7 @@ IF DIPOLES == 1:
             return ("n_replica",)
 
         def valid_keys(self):
-            return ("bjerrum_length", "prefactor", "n_replica")
+            return ("prefactor", "prefactor", "n_replica")
```
Comment (KonradBreitsprecher) [inline src/python/espressomd/magnetostatics.pyx]: prefactor appears twice in valid_keys

File: src/core/p3m.hpp [hunk 5]
```diff
@@ -125,8 +125,6 @@ extern p3m_data_struct p3m;
 
 void p3m_pre_init(void);
```
Comment (KonradBreitsprecher) [inline src/core/p3m.hpp:127]: missing void p3m_set_prefactor(void) ?

**Review by KonradBreitsprecher — COMMENTED** [review]
Pretty complete, some leftovers:
- inter.rst line 2178 Dipolar direct sum on gpu stil has bjerrum.
- prefactor appears twice in valid_keys dict in some places
- method declaration p3m_set_prefactor() in p3m.hpp
- Tutorials. I will take care of 02 charged systems. 05 and 07 also need changes

File: doc/sphinx/inter.rst [hunk 6]
```diff
@@ -1334,25 +1334,27 @@ The Coulomb (or electrostatic) interaction is defined as
...
-where :math:`l_B = e_o^2 / (4 \pi \epsilon k_B T)` denotes the Bjerrum
-length, which measures the strength of the electrostatic interaction. As
-a special case, when the thermostat is switched off, the value of
-Bjerrum length you enter is treated as :math:`l_B k_B T` rather than
-:math:`l_B`. This is used to perform an NVE integration (see also
-section :ref:`\`\`thermostat\`\`\: Setting up the thermostat`).
+where 
+
+.. math::  
+   C=\frac{1}{4pi \epsilon_0 \epsilon_r}
```
Comment (pkreissl) [inline doc/sphinx/inter.rst]: `\pi`

File: testsuite/electrostaticInteractions.py [hunk 7]
```diff
@@ -65,31 +65,72 @@ def test_p3m(self):
...
+        u_dh_core = np.zeros_like(r)
+        f_dh_core = np.zeros_like(r)
+        # need to update forces
+        for i, ri in enumerate(r):
+            self.system.part[1].pos = self.system.part[0].pos + [ri, 0, 0]
+            self.system.integrator.run(0)
+            u_dh_core[i] = self.system.analysis.energy()['coulomb']
+            f_dh_core[i] = self.system.part[0].f[0]
+
+        self.assertTrue(np.allclose(u_dh_core,
```
Comment (KaiSzuttor) [inline testsuite/electrostaticInteractions.py]: numpy.testing.assert_allclose

**Comment by RudolfWeeber** [comment]
Fixes #600

File: samples/visualization_mmm2d.py [hunk 8]
```diff
@@ -40,7 +40,11 @@
 
 system.thermostat.set_langevin(kT=0.1, gamma=1.0)
 
-mmm2d = electrostatics.MMM2D(bjerrum_length = 100.0, maxPWerror = 1e-3, const_pot = 1, pot_diff = 50.0)
+<<<<<<< HEAD
```
Comment (fweik) [inline samples/visualization_mmm2d.py]: This looks wrong.

File: src/core/interaction_data.cpp [hunk 9]
```diff
@@ -585,19 +585,30 @@ void set_dipolar_method_local(DipolarInteraction method) {
 /*                                 electrostatics */
 /********************************************************************************/
 
-int coulomb_set_bjerrum(double bjerrum) {
-  if (bjerrum < 0.0)
+int coulomb_set_prefactor(double prefactor)
+{
+  if (prefactor < 0.0) {
```
Comment (fweik) [inline src/core/interaction_data.cpp:590]: Do we really want to have that limitation? At least the gpu p3m works fine with negative prefactor (=gravity).

File: src/python/espressomd/magnetostatics.pyx [hunk 10]
```diff
@@ -34,60 +34,26 @@ IF DIPOLES == 1:
...
-            Similar to the Bjerrum length in electrostatics.
-            Gives the separation at which the dipolar interaction energy
-            corrensponds to a thermal energy of :math:`1 k_B T`.
-            If not set, an explicit choice for the `prefactor` has to be made.
-
-        prefactor : :obj:`float`, optional
-            If given, the magnetostatic prefactor is explicitly set to
-            this value. Non-optional if `bjerrum_length` is not set.
+        prefactor : :obj:`float`
+                Magnetoztatics prefactor
```
Comment (fweik) [inline src/python/espressomd/magnetostatics.pyx]: Typo

**Review by fweik — CHANGES_REQUESTED** [review]
Aside from the obvious merge leftovers, this can be merged.

**Pull request merged**